### ISSUE TITLE
How add header for request?

### ISSUE THREAD

**Issue by Myks92** [jmarioste/next-middleware-guide#1]
Please tell me how to add headers to init in [stackMiddlewares()](https://github.com/jmarioste/next-middleware-guide/blob/d561b37669551d74d1ff4317ad3f5bb887ba97b8/middlewares/stackMiddlewares.ts#L13) . For example
```js
NextResponse.next({
  request: {
     headers: //Add headers
  },
})
```
Can it be modified to add a default middleware?

**Comment by jmarioste** [comment]
There are two ways. 

```ts
export function stackMiddlewares(
  functions: MiddlewareFactory[] = [],
  index = 0
): NextMiddleware {
  const current = functions[index];
  if (current) {
    const next = stackMiddlewares(functions, index + 1);
    return current(next);
  }
  
  return () => NextResponse.next({
     request: {
        headers: //Add headers
     }
  }); 
}

```
2. Create another MiddlewareFactory that sets the default headers.
See https://github.com/jmarioste/next-middleware-guide/blob/main/middlewares/withHeaders.ts

**Comment by Myks92** [comment]
Thanks) Of course, I tried that. The first method is not suitable because there is no request. The second method for some reason does not work when sending to the server. Initially I had this and it worked:

```js
import { Next Response } from 'next/server'
import type { Next Request } from 'next/server'

export async function middleware(request: Next Request) {
  const token = request.cookies.get('next-auth.session-token')?.value
  const requestHeaders = new Headers(request.headers)
  request Headers.set('Authorization', 'Bearer ' + token)

  const response = Next Response.next({
    request: {
      headers: requestHeaders,
    },
  })
  return response
}

export const config = {
  matcher: [‘/api/site/:path*'],
}
```

**Comment by jmarioste** [comment]
Do you have any other middlewares aside from this? If you have middlewares that cut-off the chain, perhaps you're already sending a different `NextResponse`.

If you need to keep the response you can do something like this
```ts
import { NextFetchEvent, NextMiddleware, NextRequest } from "next/server";

import { MiddlewareFactory } from "./types";

export const withAuthorizationHeader: MiddlewareFactory = (next) => {
  return async (request: NextRequest, _next: NextFetchEvent) => {
    const token = request.cookies.get("next-auth.session-token")?.value;

    request.headers.set("Authorization", "Bearer " + token);
    const res = next(request, _next); //get the response from the next middleware
    return res;
  };
};

```

**Comment by jmarioste** [comment]
Hi @Myks92,I just forgot. In the first example you can actually access the request because it returns a middleware function. Here's how to do it:

```tsx
import { NextMiddleware, NextRequest, NextResponse } from 'next/server'
import { MiddlewareFactory } from './types'

/**
 * A utility to create middle ware functions like this middleware1(middleware2(middlware3)).
 * @returns
 */
export function stackMiddlewares(
  functions: MiddlewareFactory[] = [],
  index = 0
): NextMiddleware {
  const current = functions[index]
  if (current) {
    const next = stackMiddlewares(functions, index + 1)
    return current(next)
  }
  const defaultMiddleware = (request: NextRequest) => {
    const token = request.cookies.get('next-auth.session-token')?.value
    const requestHeaders = new Headers(request.headers)
    requestHeaders.set('Authorization', 'Bearer ' + token)

    return NextResponse.next({
      request: {
        headers: requestHeaders
      }
    })
  }
  return defaultMiddleware
}

```

Hope this helps.

**Comment by Myks92** [comment]
Yes, thank you! The first example I got is exactly the same as yours. Even the name is the same)

The second example thought that it was possible to do without adding `defaultMiddleware` to `stackMiddlewares`, but apparently not provided. Thanks again. If possible, think about extract `defaultMiddleware`  from `stackMiddlewares`. By the way, I named `stackMiddlewares` as `chainMiddlewares`. I think this is a more appropriate name.

Thanks your help! Good realisation!
